### DISCUSSION TITLE
Caching compiled rego

### DISCUSSION THREAD

**Discussion by dushyant7917** [open-policy-agent/community#285] [OPA and Rego]
Write now eval command compiles and executes the rego files. Majority of the time is taken by compilation step. Its a bottleneck if the policy is heavy or complex in nature. Is there any plan in the near future to cache the compiled rego so that whenever an eval request comes in OPA server, it can use the cached compiled rego and just perform the execution step to give the result.

**Comment by srenatus** [comment]
I'm confused about some things here. The `opa eval` CLI works differently from the server -- the server won't compile all policies on an incoming request on the Data API. It'll only compile all policies when a policy changes. It'll also cache compiled queries.

What is it that you're seeing that makes you think the server did compile policies for every query?

**Reply by dushyant7917** [reply]
Sorry for the confusion @srenatus. I was under the impression that OPA server also uses OPA cli under the hood so caching of compiled policies won't be happening. In that case I think OPA server can't do any more optimization? Is using WASM policies over Rego policies a good idea? Theoretically it looks appealing but I have seen issues ([link1](https://github.com/open-policy-agent/opa/issues/4704), [link2](https://github.com/open-policy-agent/opa/issues/2281)) where people have mentioned that it's not performing as expected compared to Rego.

**Reply by srenatus** [reply]
> In that case I think OPA server can't do any more optimization? Is using WASM policies over Rego policies a good idea? Theoretically it looks appealing but I have seen issues (https://github.com/open-policy-agent/opa/issues/4704, https://github.com/open-policy-agent/opa/issues/2281) where people have mentioned that it's not performing as expected compared to Rego.

✔️ OPA server optimizes as much as it can, I think.

On Wasm, I'd advise an experimental approach -- Try and see if it performs better for what you need, and evaluate based on that. It can beat Rego, but it won't on each and every occasion, since the Rego evaluator is quite an optimized beast as well.

**Comment by ramaraochavali** [comment]
> the server won't compile all policies on an incoming request on the Data API. It'll only compile all policies when a policy changes. It'll also cache compiled queries.

Just to clarify, OPA server does not compile a rego file on every request but only compiles when a policy is changed. Is that right?
